cpp/dolfin/io/HDF5Interface.cpp

```diff
@@ -34,37 +33,45 @@ hid_t HDF5Interface::open_file(MPI_Comm mpi_comm, const std::string filename,
     assert(status != HDF5_FAIL);
     MPI_Info_free(&info);
 #else
-    throw std::runtime_error("Cannot use MPI-IO output if DOLFIN is not configured with MPI");
+    throw std::runtime_error(
+        "Cannot use MPI-IO output if DOLFIN is not configured with MPI");
 #endif
   }
 #endif
 
+  // Query file existence only once
+  bool file_exists = boost::filesystem::exists(filename);
+
   hid_t file_id = HDF5_FAIL;
   if (mode == "w")
   {
     // Create file for write, (overwriting existing file, if present)
     file_id = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, plist_id);
   }
-  else
+  else if (mode == "a")
   {
-    // Check that file exists
-    if (!boost::filesystem::is_regular_file(filename))
+    if (file_exists)
     {
-      log::dolfin_error("HDF5Interface.cpp", "open HDF5 file",
-                   "File \"%s\" does not exist", filename.c_str());
+      file_id = H5Fopen(filename.c_str(), H5F_ACC_RDWR, plist_id);
     }
-
-    if (mode == "a")
+    else
+    {
+      // If file doesn't exists but we want to append it
+      // create new
+      file_id
+          = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, plist_id);
+    }
+  }
+  else if (mode == "r")
+  {
+    if (file_exists)
     {
-      // Open file existing file for append
       file_id = H5Fopen(filename.c_str(), H5F_ACC_RDWR, plist_id);
     }
-    else if (mode == "r")
-      file_id = H5Fopen(filename.c_str(), H5F_ACC_RDONLY, plist_id);
     else
     {
       log::dolfin_error("HDF5Interface.cpp", "open HDF5 file",
```

Comment:
Could you change this to `throw std::runtime_error("...")`? (Shame we can't use `std::filesystem` yet . . . )